app/routes.py

```diff
@@ -1,2 +1,291 @@
+            "description": task.description,
+            "is_complete": True
+        }
+    }, 200
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_incomplete(task_id):
+    task = Task.query.get(task_id)
+
+    if task == None:
+        return make_response("Task {taske_id} not found", 404)
+    task.completed_at = None
+
+    db.session.commit()
+
+    return {
+        "task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete": False
+        }
+    }, 200
+
+@goals_bp.route("", methods=["POST", "GET"])
+def handle_goals():
+    if request.method == "POST":
+        request_body = request.get_json()
+        if "title" not in request_body:
+            return {
+                "details": f"Invalid data"
+            },400
+        new_goal = Goal(title=request_body["title"])
+        db.session.add(new_goal)
+        db.session.commit()
+
+        return make_response({
+            "goal": {
+                "id": new_goal.goal_id,
+                "title": new_goal.title
+
+            }
+        }, 201)
+    if request.method == "GET":
+        goals = Goal.query.all()
+        goals_response = []
+        for goal in goals:
+            goals_response.append({
+                "id": goal.goal_id,
+                "title": goal.title
+            })
+        return make_response(jsonify(goals_response))
+
+@goals_bp.route("/<goal_id>", methods=["GET", "PUT", "DELETE"])
+def handle_goal(goal_id):
+    goal = Goal.query.get(goal_id)
+    if goal is None:
+        return make_response(f"", 404)
+
+    if request.method == "GET":
+        return make_response({
+            "goal": {
+                "id": goal.goal_id,
+                "title": goal.title
+            }
+        })
+    elif request.method == "PUT":
+        form_data = request.get_json()
+        goal.title = form_data["title"]
+        db.session.commit()
+        goal_response = {
+            "goal": {
+                "id": goal.goal_id,
+                "title": goal.title
+            }
+        }
+        return make_response(goal_response, 200)
+
+    elif request.method == "DELETE":
+        db.session.delete(goal)
+        db.session.commit()
+
+        return make_response(
+            {
+                "details": f"Goal {goal.goal_id} \"{goal.title}\" successfully deleted"
+            }
+        )
+        
+@goals_bp.route("/<goal_id>/tasks", methods=["GET", "POST"])
+def handle_goals_and_tasks(goal_id):
+    if request.method == "POST":
+        goal = Goal.query.get(goal_id)
+        if not goal:
+            return make_response({
+                f"Goal # {goal_id} not found."
+            }, 404)
+        request_body = request.get_json()
+        for id in request_body["task_ids"]:
+            task = Task.query.get(id)
+            goal.tasks.append(task)
+            db.session.add(goal)
+            db.session.commit()
```

Comment:
👍 

We do need to retrieve each `Task` to add it to the goal using the SqlAlchemy-created `tasks` field. But since the `goal` and each `task` are already retrieved from the DB, we don't need to add them to the `session`. We should also be sure to validate that the `Task` retrieval succeeded.

We can also wait to do the commit until after adding all the tasks. This will have the effect of committing this change all together in a single transaction, rather than running the risk of some of the tasks being added, and then possibly running into an error part of the way through (again, what if one of the task ids is invalid?).

Also, what would happen if the goal previously had some tasks set. Do we want to add the new tasks to the existing tasks? Do we want to replace them and sever any prior task → goal relationships? What behavior is implemented here?